research/fed-bpt/src/cma_decomposer.py

```diff
@@ -0,0 +1,116 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+from typing import Any, Type
+
+import cma
+import numpy as np
+from cma import CMADataLogger, CMAOptions
+from cma.constraints_handler import BoundNone
+from cma.evolution_strategy import _CMAParameters, _CMASolutionDict_functional, _CMAStopDict
+from cma.optimization_tools import BestSolution
+from cma.recombination_weights import RecombinationWeights
+from cma.sampler import GaussFullSampler
+from cma.sigma_adaptation import CMAAdaptSigmaCSA
+from cma.transformations import DiagonalDecoding, GenoPheno
+from cma.utilities.utils import BlancClass, DictFromTagsInString, ElapsedWCTime, MoreToWrite, SolutionDict
+
+from nvflare.app_common.decomposers.common_decomposers import Float64ScalarDecomposer, NumpyArrayDecomposer
+from nvflare.fuel.utils import fobs
+from nvflare.fuel.utils.fobs import Decomposer
+from nvflare.fuel.utils.fobs.datum import DatumManager
+
+
+class GaussFullSamplerDecomposer(Decomposer):
+    def supported_type(self) -> Type[GaussFullSampler]:
+        return GaussFullSampler
+
+    def decompose(self, target: GaussFullSampler, manager: DatumManager = None) -> Any:
+        target = copy.deepcopy(target)
+        members = vars(target)
+        # The functions can't be serialized
+        if "randn" in members:
```

Comment:
maybe a configurable list of "unserializable items" rather than one by one?

Reply:
Currently, this is not a general purpose decomposer. So it should be fine.